pkg/gameservers/health_test.go

```diff
@@ -79,15 +79,114 @@ func TestHealthUnschedulableWithNoFreePorts(t *testing.T) {
+				}}
+			}, expected: true},
+		"after ready and terminated container": {
+			setup: func(gs *agonesv1.GameServer, pod *corev1.Pod) {
+				gs.Status.State = agonesv1.GameServerStateReady
+				pod.Status.ContainerStatuses = []corev1.ContainerStatus{{
+					Name:  gs.Spec.Container,
+					State: corev1.ContainerState{Terminated: &corev1.ContainerStateTerminated{}},
+				}}
+			},
+			expected: false,
+		},
+		"before ready, with no terminated container": {
+			setup: func(gs *agonesv1.GameServer, pod *corev1.Pod) {
+				gs.Status.State = agonesv1.GameServerStateScheduled
+			},
+			expected: false,
+		},
+		"after ready, with no terminated container": {
+			setup: func(gs *agonesv1.GameServer, pod *corev1.Pod) {
+				gs.Status.State = agonesv1.GameServerStateAllocated
+			},
+			expected: false,
+		},
+		"before ready, with a LastTerminated container": {
+			setup: func(gs *agonesv1.GameServer, pod *corev1.Pod) {
+				gs.Status.State = agonesv1.GameServerStateScheduled
+				pod.Status.ContainerStatuses = []corev1.ContainerStatus{{
+					Name:                 gs.Spec.Container,
+					LastTerminationState: corev1.ContainerState{Terminated: &corev1.ContainerStateTerminated{}},
+				}}
+			},
+			expected: true,
+		},
+		"after ready, with a LastTerminated container, not matching": {
+			setup: func(gs *agonesv1.GameServer, pod *corev1.Pod) {
+				gs.Status.State = agonesv1.GameServerStateAllocated
+				gs.Annotations[agonesv1.GameServerReadyContainerIDAnnotation] = "4321"
+				pod.Status.ContainerStatuses = []corev1.ContainerStatus{{
+					ContainerID:          "1234",
+					Name:                 gs.Spec.Container,
+					LastTerminationState: corev1.ContainerState{Terminated: &corev1.ContainerStateTerminated{}},
+				}}
+			},
+			expected: false,
+		},
+		"after ready, with a LastTerminated container, matching": {
+			setup: func(gs *agonesv1.GameServer, pod *corev1.Pod) {
+				gs.Status.State = agonesv1.GameServerStateAllocated
+				gs.Annotations[agonesv1.GameServerReadyContainerIDAnnotation] = "1234"
+				pod.Status.ContainerStatuses = []corev1.ContainerStatus{{
+					ContainerID:          "1234",
+					Name:                 gs.Spec.Container,
+					LastTerminationState: corev1.ContainerState{Terminated: &corev1.ContainerStateTerminated{}},
+				}}
+			},
+			expected: true,
+		},
+	}
+
+	for k, v := range fixtures {
+		t.Run(k, func(t *testing.T) {
+			m := agtesting.NewMocks()
+			hc := NewHealthController(healthcheck.NewHandler(), m.KubeClient, m.AgonesClient, m.KubeInformerFactory, m.AgonesInformerFactory)
+			gs := &agonesv1.GameServer{ObjectMeta: metav1.ObjectMeta{Name: "test", Namespace: defaultNs}, Spec: newSingleContainerSpec()}
+			gs.ApplyDefaults()
+			pod, err := gs.Pod()
+			assert.NoError(t, err)
+
+			v.setup(gs, pod)
+
+			m.KubeClient.AddReactor("list", "pods", func(action k8stesting.Action) (bool, runtime.Object, error) {
+				return true, &corev1.PodList{Items: []corev1.Pod{*pod}}, nil
```

Comment:
It looks like this doesn't cover the branches where pod doesn't exist or k8s returns an error to put the request back into the queue.